File: src/Samples/FSharpApp/App.ViewModel/CustomGraphObjects.fs
namespace SampleApp.ViewModel

open GraphSharp.Controls
open QuickGraph
open SampleApp.Model
open System
open System.Collections.Generic
open System.Diagnostics
open Utopia
open Utopia.ViewModel

[<DebuggerDisplay("{ID}-{IsInput}")>]
type CustomVertex(node : INode) as this = 
    inherit ViewModelBase()
    
    let tryParseInt(v : string) = 
        try 
            Convert.ToInt32(v) |> Some
        with _ -> None
    
    do 
        node.Changed.Add(fun args -> 
            this.OnPropertyChanged "Value"
            this.OnPropertyChanged "Dirty")
    
    member this.IsValidValue() = 
        let isInt(v : string) = 
            try 
                Convert.ToInt32(v) |> ignore
                true
            with _ -> false
        if String.IsNullOrEmpty(this.Value) || this.Value = null then "missing value"
        elif tryParseInt(this.Value).IsNone then "not a valid number"
        else null
    
    override this.ValidatedProperties = [| "Value" |]
    
    override this.GetValidationError(propertyName) = 
        match propertyName with
        | "Value" -> this.IsValidValue()
        | _ -> 
            Debug.Fail("Unexpected property being validated on Vertex: " + node.ID)
            Unchecked.defaultof<string>
    
    member val Node = node
    member val ID = node.ID
    member val IsInput = (node.DependentNodes.Length = 0)
    
    member this.Value 
        with get () = string node.UntypedValue
        and set v = 
            match tryParseInt(v) with
            | Some x -> node.UntypedValue <- box x
            | None -> ()
    
    member this.Dirty with get () = node.Dirty
    override this.ToString() = sprintf "{%s}-{%b}" this.ID this.IsInput

[<DebuggerDisplay("{source.ID} -> {target.ID}")>]
type CustomEdge(source, target) = 
    inherit Edge<CustomVertex>(source, target)
    member this.ID : string = sprintf "{%s} -> {%s}" source.ID target.ID
    override this.ToString() = this.ID

type CustomGraph(ce : ICalculationEngine) as this = 
    inherit BidirectionalGraph<CustomVertex, CustomEdge>()
    let vertices = new Dictionary<string, CustomVertex>()
    
    do 
        for n in ce.Nodes do
            let vertex = new CustomVertex(n)
            vertices.Add(n.ID, vertex)
            this.AddVertex(vertex) |> ignore
        for n in ce.Nodes do
            for d in n.DependentNodes do
                let edge = CustomEdge(vertices.[d.ID], vertices.[n.ID])
                this.AddEdge(edge) |> ignore
    
    member this.CalcEngine = ce
    member this.UpdateNode(id : string) = 
        let vertex = vertices.[id]
        vertex.Node.AsyncCalculate()

type CustomGraphLayout() = 
    inherit GraphLayout<CustomVertex, CustomEdge, CustomGraph>()

File: src/Samples/CSharpApp/App.ViewModel/CustomGraphObjects.cs
namespace CSharpApp.ViewModel
{
    using System;
    using System.Collections.Generic;
    using GraphSharp.Controls;
    using QuickGraph;
    using Utopia;
    using Utopia.ViewModel;

    public class CustomVertex : ViewModelBase
    {
        public CustomVertex(INode node)
            : base()
        {
            Node = node;

            Node.Changed += new ChangedEventHandler(OnNodeChanged);
        }

        void OnNodeChanged(object sender, EventArgs args)
        {
            this.OnPropertyChanged("Value");
            this.OnPropertyChanged("Dirty");
        }

        public INode Node { get; private set; }

        public string ID { get { return Node.ID; } }

        public bool IsInput { get { return Node.DependentNodes.Length == 0; } }

        public string Value
        {
            get
            {
                if (Node.UntypedValue != null) { return Node.UntypedValue.ToString(); }
                return null;
            }
            set
            {
                int result = 0;
                try
                {
                    result = Convert.ToInt32(value);
                    Node.UntypedValue = (object)result;
                }
                catch (Exception) { }
            }
        }

        public bool Dirty { get { return Node.Dirty; } }

        public override string ToString() { return string.Format("{0}-{1}", ID, IsInput); }
    }

    public class CustomEdge : Edge<CustomVertex>
    {
        public string ID { get; private set; }

        public CustomEdge(CustomVertex source, CustomVertex target)
            : base(source, target)
        {
            ID = string.Format("{0} -> {1}", source.ID, target.ID);

        }

        public override string ToString()
        {
            return ID;
        }
    }

    public class CustomGraph : BidirectionalGraph<CustomVertex, CustomEdge>
    {
        IDictionary<string, CustomVertex> _vertices;

        public CustomGraph(ICalculationEngine calculationEngine)
            : base()
        {
            _vertices = new Dictionary<string, CustomVertex>();

            foreach (var node in calculationEngine.Nodes)
            {
                var vertex = new CustomVertex(node);
                _vertices.Add(node.ID, vertex);
                AddVertex(vertex);
            }

            foreach (var node in calculationEngine.Nodes)
            {
                foreach (var dependentNode in node.DependentNodes)
                {
                    var edge = new CustomEdge(_vertices[dependentNode.ID], _vertices[node.ID]);
                    AddEdge(edge);
                }
            }
        }

        public void Update(string id)
        {
            var vertex = _vertices[id];
            vertex.Node.AsyncCalculate();
        }

    }

    public class CustomGraphLayout : GraphLayout<CustomVertex, CustomEdge, CustomGraph>
    {
        public CustomGraphLayout()
            : base()
        {
        }
    }
}